### DISCUSSION TITLE
Inverted colours on HyperHDR

### DISCUSSION THREAD

**Discussion by rama31244** [awawa-dev/HyperHDR#867] [Q&A]
Hi, i run HyperHDR 20.0.0 on a raspberry pi 3b on the latest version of raspberry pi OS and it's been working fine for a long time but today the colours were all inverted when I was watching movies. I couldn't find a setting to invert colours, how can I troubleshoot this?
Thanks

**Comment by awawa-dev** [comment]
Hi
Please read the pinned topic https://github.com/awawa-dev/HyperHDR/discussions/848 maybe your system got updated.

**Comment by rama31244** [comment]
Thanks, should have checked the pins first, my bad

**Reply by awawa-dev** [reply]
No problem, if you decide to try "rpi-update" (and restart after that), let me know if that solves the issue. And what version of Kernel it will install now. Which encoding are you using for your grabber (yuv/mjpeg etc)?

**Reply by rama31244** [reply]
I'm trying the rpi-update now. How do i check which encoding i'm using?

**Reply by awawa-dev** [reply]
On the main page:
![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/4dc2d094-16f5-4a73-ad50-99358f91ac3a)

**Comment by rama31244** [comment]
Video mode: 800x600x60 MJPEG
And the rpi-update didn't fix the problem. What is the easiest way to downgrade the kernel? Sorry, i'm fairly new to linux so still learning

**Reply by awawa-dev** [reply]
One more thing. Do you have tone mapping enabled? If so, could you disable it?

**Comment by rama31244** [comment]
If you're referring to the HDR tone mapping in the Remote settings menu, it is set to off

**Comment by awawa-dev** [comment]
So it's probably https://github.com/libjpeg-turbo/libjpeg-turbo library issue/bug on latest kernel, because HyperHDR doesn't interfere with its stream when the tone mapping is disabled. If your grabber supports YUV encoding then you can give a try, then we will know if HyperHDR's own YUV processing is affected also.

I havent test it so it's better to make a backup of HyperHDR configuration first (general tab in the UI):
According to https://github.com/raspberrypi/rpi-update?tab=readme-ov-file#options
`sudo rpi-update d86b5843d68b9972a5430a6d3da1b271cfc83521`

This hash is for Kernel 6.1.74 https://github.com/raspberrypi/rpi-firmware/commit/d86b5843d68b9972a5430a6d3da1b271cfc83521

**Comment by rama31244** [comment]
Sorry for all the dumb questions but how do I turn on YUV encoding? I tried to find it in settings on the video capture page but it's not there

**Reply by awawa-dev** [reply]
This depends on whether your grabber supports YUV (sometimes it is only available when the grabber is connected to a USB 3.0 port, but this is not the rule). Look here https://github.com/awawa-dev/HyperHDR/discussions/644 click the yellow "info" button in the second screenshot and you will see the available modes.

**Reply by awawa-dev** [reply]
or here (wiki): https://github.com/awawa-dev/HyperHDR/wiki/Quick-start#configuring-video--system-grabber

**Reply by rama31244** [reply]
I tried YUYV encoding and it was still inverted. Downgrading kernel seems to be the only way to get it to work at the moment

**Reply by awawa-dev** [reply]
@rama31244 because now I have a lot o free time I tested it and is it a joke? everything works fine on kernel 6.6.20 and latest 6.6.31
![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/2aa6710d-2ae6-4da3-b557-1a46cf912971)
![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/2143cc01-15e1-4ebc-a3cf-20b70bf8301d)
![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/51a95b8d-f23c-46ac-939c-b1490dc9d979)

both yuv and mjpeg work OK

**Reply by rama31244** [reply]
Hmm, that's so odd. I tried the newest kernel last night with sudo rpi-update and the colors were inverted. I run a pi 3b so not sure if that makes a difference

**Reply by awawa-dev** [reply]
Which OS version? Legacy arm6l or aarch64?

**Reply by awawa-dev** [reply]
And is it really Raspberry Pi OS lite or some other Debian variant?

**Reply by rama31244** [reply]
aarch64 and raspberry pi OS lite (latest bookworm version)

**Reply by awawa-dev** [reply]
what grabber are you using? Please paste the results of `vcgencmd version` and `uname -r`

**Comment by rama31244** [comment]
Grabber: Rullz 4k 60hz USB3.0 Loop 
uname -r: 6.1.74-v8+
vcgencmd: version 30f0c5e4d076da3ab4f341d88e7d505760b93ad7 (clean) (release) (start)

This is after i've downgraded the kernel obviously

**Reply by awawa-dev** [reply]
Thanks. I changed the same card from RPi4 to RPi3 and indeed the colors became broken. I think this issue (byte order) may be the source of many of the issues affecting RPi3 on Kernel 6.6. I haven't found any issue reported to the Raspberry Pi repository, so I think that since the RPi4 and possibly the RPi5 are working properly, it seems that the A53 architecture is also outdated for them and users and they are giving it limited attention in testing. You need to slowly get used to this idea and think about switching to RPi4/5 also because of USB3.0.

Anyway, if you want to have this fixed for A53 on kernel 6.6, please report it to Raspberry Pi here on Github or on their forum in the appropriate section. Otherwise nothing will change.

**Comment by rama31244** [comment]
Ok thanks for all your help, the pi is attached behind my wall mounted 85inch tv so I've been dreading the change but it sounds like I need to do it sooner or later. Do I lose anything by continuing to run the latest version of hyoerhdr with a downgraded kernel?
